### DISCUSSION TITLE
Swift bindings improvement

### DISCUSSION THREAD

**Discussion by aimore** [dotnet/maui#59] [General]
## Summary
1-Improve documentation and have more articles/tutorials for Native bindings.
2-Although we have a "proxy" solution for Swift bindings, it still a painful process specially for complex Frameworks with several dependencies.
3-Since we have Objective Sharpie tool, would be great to have something similar for Swift Bindings.

Imagine a world where companies don't need to worry about migrating to cross-platform because they know that all native Libraries/Frameworks they depend on can be reused through bindings!

**Comment by dhindrik** [comment]
I think this is an issue that belongs to https://github.com/xamarin/xamarin-macios because it is more about the underlying platform.

**Comment by davidortinau** [comment]
https://github.com/xamarin/binding-tools-for-swift

**Reply by noobsoftware** [reply]
Will this tool be updated? It is currently incompatible with Xcode.

**Comment by aimore** [comment]
@dhindrik  I agree with that :)
@davidortinau  thanks for the link, but this is far from what normally companies are looking for... when it comes to a full solution for migrating/binding native iOS frameworks.
ref: https://github.com/xamarin/binding-tools-for-swift/issues/215
There are a whole lot of things that can be done/improved, I have some suggestions and few things that could be already implemented to facilitate this process. 
But maybe this is not the right place

**Comment by GiampaoloGabba** [comment]
I think that the whole "bindings to libraries" system is the most underrated feature ever :)
As far as i know, Xamarin is the only crossplatform framework  that can bind to any existing native SDK fro iOS and Android.

But the whole process is difficult, requires a lot of times, there are fragmented informations (not always up to date).... so a lot of people is not even considering it.

It would be awesome if microsoft can develop a gui tool where you can select your Android/iOS SDKs (whether they are made in objC, swift, Java, Kotlin) and produces automatic bindings with a click and then starts a wizard where we can select the methods we want, propose the renaming with the proper casing, ecc... After that we have our binding projects done that we can customize by hand.

This would be glorious and will encourage developers to use it. I repeat, this is a HUGE advantage to Xamarin as a crossplatform framework that other does not have, so lets take advantage of it as much as we can!

**Reply by fergus-repo** [reply]
I agreed, it's too difficult to make the binding. So I switch to Kotlin KMM, I hope MS will provide such tool and I will come back.

**Comment by softlion** [comment]
Hello mates :)
Any news on that ?

This limitation is starting to make harder the implementation of some common features:
- StoreKit2 (swift only): can not be used, and is required to put a "userId" metadata in the purchase ticket.
- RealityKit (swift only): simplify 3D display. SceneKit is not evolving anymore.
- App extracts, Home screen Widgets must be written in Swift

And probably much more.

There is the same issue on Android for libraries not written in java.

That makes Xamarin/maui less shiny.
